### ISSUE TITLE
Entity Framework 7 Support, IQueryable Extensions

### ISSUE THREAD

**Issue by freudj76** [eswann/Mapster#27]
Hi,

First of all, thank you for such a great library. Just switched from AutoMapper to Mapster, and I think it's awesome. 
I have a question regarding the IQueryable extensions using EF7. When I call the Project() method where the underlying ORM is EF7, I am getting an error as follows

    Expression of type 'Microsoft.Data.Entity.Query.Internal.EntityQueryable' cannot be used for
    parameter of type 'System.Collections.Generic.ICollection`' of method.

Is this a problem because of EF7, or possibly an issue within Mapster? Any ideas? Will be happy to provide more information if necessary.

Thanks!

**Comment by eswann** [comment]
Thanks for the heads-up...

**Comment by chaowlert** [comment]
Could you provide source class and destination class for inspection? And please also provide code when you call projection.

**Comment by freudj76** [comment]
Source classes
```
    public class Product
    {
        public int AvailableQuantity { get; set; }
        public string Brand { get; set; }
        public int Cost { get; set; }
        public string Description { get; set; }
        public bool HasVariants { get; set; }
        public string ImageUrl { get; set; }
        public int InStockQuantity { get; set; }
        public bool IsActive { get; set; }
        public int LowInventoryQuantity { get; set; }
        public string Name { get; set; }
        public int Id { get; set; }
        public virtual ProductType ProductType { get; set; }

        public int ProductTypeId { get; set; }
        public int RetailPrice { get; set; }
        public string Sku { get; set; }
        public Supplier Supplier { get; set; }

        public virtual Company Company { get; set; }
        public int CompanyId { get; set; }

        public string SupplierCode { get; set; }
        public int SupplierId { get; set; }
        public ICollection<ProductTag> Tags { get; set; }
        public int WholesalePrice { get; set; }
        public virtual ICollection<ProductCustomField> CustomFields { get; set; }
        public virtual ICollection<ProductPricePlan> PricePlans { get; set; }
    }

public class CustomField
    {
        public virtual ProductType ProductType { get; set; }
        public int ProductTypeId { get; set; }
        public int Id { get; set; }
        public string DisplayName { get; set; }
        public string Name { get; set; }
    }

public class PricePlan
    {
        public int CompanyId { get; set; }
        public string DisplayName { get; set; }
        public string Name { get; set; }
        public int Id { get; set; }

        public Company Company { get; set; }
    }
```

Destination classes

```
public class ProductBindingModel
        {
            public int AvailableQuantity { get; set; }
            public string Brand { get; set; }
            public int Cost { get; set; }
            public string Description { get; set; }
            public bool HasVariants { get; set; }
            public string ImageUrl { get; set; }
            public int InStockQuantity { get; set; }
            public bool IsActive { get; set; }
            public int LowInventoryQuantity { get; set; }
            public string Name { get; set; }
            public int Id { get; set; }
            public int ProductTypeId { get; set; }
            public int RetailPrice { get; set; }
            public string Sku { get; set; }
            public int CompanyId { get; set; }
            public string SupplierCode { get; set; }
            public int SupplierId { get; set; }
            public int WholesalePrice { get; set; }
            public List<ProductCustomFieldBindingModel> CustomFields { get; set; }
            public List<ProductPricePlanBindingModel> PricePlans { get; set; }

        }

public class ProductCustomFieldBindingModel
        {
            public int Id { get; set; }
            public int ProductId { get; set; }
            public int CustomFieldId { get; set; }
            public string Value { get; set; }
        }

        public class ProductPricePlanBindingModel
        {
            public int Id { get; set; }
            public int ProductId { get; set; }
            public int PricePlanId { get; set; }
            public string Value { get; set; }
        }
```

Mapping Config
```
TypeAdapterConfig<Data.Entities.Product, List.ProductBindingModel>.NewConfig();
TypeAdapterConfig<Data.Entities.ProductType, ProductTypeBindingModel>.NewConfig();
 TypeAdapterConfig<Data.Entities.CustomField, CustomFieldBindingModel>.NewConfig();
```
Where projection is called
```
var products = _context.Products
                    .Include(p => p.CustomFields)
                    .Include(p => p.PricePlans)
                    .Where(p => p.CompanyId == message.CompanyId)
                    .Project().To<ProductBindingModel>()
                    .ToPagedList(pageNumber, pageSize);
```

Let me know if you need anything else. I appreciate the lightning fast response.

**Comment by chaowlert** [comment]
My guess is Mapster cannot map `List<>` in `ProductBindingModel`. Projection does not know how to create list in sql (even EF will create list for collection internally). 

I will find the way to map to `List<>` in projection. In the mean time, could you change `List<>` to `IEnumerable<>`?

**Comment by chaowlert** [comment]
PS. You do not need to declare mapping config 

```
TypeAdapterConfig<Data.Entities.Product, List.ProductBindingModel>.NewConfig();
TypeAdapterConfig<Data.Entities.ProductType, ProductTypeBindingModel>.NewConfig();
 TypeAdapterConfig<Data.Entities.CustomField, CustomFieldBindingModel>.NewConfig();
```

Mapster automatically create config unless you set `RequireExplicitMapping` to true.

**Comment by freudj76** [comment]
`IEnumerable<>` worked like a champ. Really appreciate the help. This library is awesome

**Comment by eswann** [comment]
Per pull request #30, this should now be closed and functional with release 2.1.0.
